### PR TITLE
CORE-11847: Populate 404 correctly for membership APIs

### PR DESCRIPTION
**Pull request by yift-r3**
Populate a specific kind of persistent error for not found and conflict issues from the DB worker to the REST worker, and make the MGM resource return the correct error in those cases.

API pull request https://github.com/corda/corda-api/pull/1576

### CODE REVIEW

**Comment by yift-r3** [comment]
build e2e

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt [hunk 1]
```diff
@@ -101,15 +102,15 @@ internal class AddNotaryToGroupParametersHandler(
...
                 "Cannot add notary to group parameters, no group parameters found."
             )
         }
 
         val parametersMap = deserializer.deserializeKeyValuePairList(previous.singleResult.parameters).toMap()
         val notaryInfo = memberInfoFactory.createMemberInfo(notaryMemberInfo)
         val notary = notaryInfo.notaryDetails
-            ?: throw MembershipPersistenceException(
+            ?: throw NotFoundEntityPersistenceException(
                 "Cannot add notary to group parameters - notary details not found."
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt]: maybe we could add a holding id or name?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddPreAuthTokenHandler.kt [hunk 2]
```diff
@@ -4,23 +4,29 @@ import net.corda.data.membership.db.request.MembershipRequestContext
...
+                        request.ttl,
+                        PreAuthTokenStatus.AVAILABLE.toString(),
+                        creationRemark = request.remark,
+                        removalRemark = null
+                    )
                 )
-            )
+            }
+        } catch (_: EntityExistsException) {
+            throw ConflictPersistenceException("Token with ID: ${request.tokenId} already exists.")
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddPreAuthTokenHandler.kt]:
```suggestion
            throw ConflictPersistenceException("Token with ID: '${request.tokenId}' already exists.")
```

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt [hunk 3]
```diff
@@ -101,15 +102,15 @@ internal class AddNotaryToGroupParametersHandler(
             .setLockMode(LockModeType.PESSIMISTIC_WRITE)
             .setMaxResults(1)
         if (previous.resultList.isEmpty()) {
-            throw MembershipPersistenceException(
+            throw NotFoundEntityPersistenceException(
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/AddNotaryToGroupParametersHandler.kt:105]: maybe we could add a holding id or name?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistApprovalRuleHandler.kt [hunk 4]
```diff
@@ -27,7 +27,7 @@ internal class PersistApprovalRuleHandler(
                 )
             if (em.createQuery(query).resultList.isNotEmpty()) {
                 logger.warn("Approval rule not added as an identical rule already exists.")
-                throw MembershipPersistenceException("Approval rule not added as an identical rule already exists.")
+                throw ConflictPersistenceException("Approval rule not added as an identical rule already exists.")
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistApprovalRuleHandler.kt]: add id?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistMemberInfoHandler.kt [hunk 5]
```diff
@@ -59,15 +60,15 @@ internal class PersistMemberInfoHandler(
                         val updatedMemberContext = deserialize(it.signedMemberContext.data.array())
                         val updatedMGMContext = deserialize(it.serializedMgmContext.array())
                         if (currentMemberContext.items != updatedMemberContext.items) {
-                            throw MembershipPersistenceException(
+                            throw ConflictPersistenceException(
                                 "Cannot update member info with same serial number " +
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistMemberInfoHandler.kt:64]: add name/holding id?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistMemberInfoHandler.kt [hunk 6]
```diff
@@ -59,15 +60,15 @@ internal class PersistMemberInfoHandler(
...
                                     "(${newMemberInfo.serial}): member context differs from original."
                             )
                         }
                         if (currentMgmContext.toMap().removeTime() != updatedMGMContext.toMap()
                                 .removeTime()
                         ) {
-                            throw MembershipPersistenceException(
+                            throw ConflictPersistenceException(
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/PersistMemberInfoHandler.kt:71]: add name/holding id?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt [hunk 7]
```diff
@@ -132,7 +133,7 @@ internal class SuspendMemberHandler(
             .setLockMode(LockModeType.PESSIMISTIC_WRITE)
             .setMaxResults(1)
         if (previous.resultList.isEmpty()) {
-            throw MembershipPersistenceException(
+            throw NotFoundEntityPersistenceException(
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:136]: add name/holding id?

File: components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt [hunk 8]
```diff
@@ -141,7 +142,7 @@ internal class SuspendMemberHandler(
             keyValuePairListDeserializer.deserializeKeyValuePairList(previous.singleResult.parameters).toMap()
         val notaryInfo = memberInfoFactory.createMemberInfo(memberInfo)
         val notary = notaryInfo.notaryDetails
-            ?: throw MembershipPersistenceException(
+            ?: throw NotFoundEntityPersistenceException(
```
Comment (nikinagy) [inline components/membership/membership-persistence-service-impl/src/main/kotlin/net/corda/membership/impl/persistence/service/handler/SuspendMemberHandler.kt:145]: add name/holding id?

**Review by nikinagy — COMMENTED** [review]
I think if we refactor the exception throwing then maybe we could add a bit more details to these messages to ensure easier investigation on failures.

**Review by nikinagy — DISMISSED** [review]
LGTM

**Review by nikinagy — APPROVED** [review]
LGTM

**Pull request merged**